### PR TITLE
Adding support for mysql upsert via AR import

### PR DESCRIPTION
**Pull request by mitch-lindsay**
The existing scrubbing process relies on singular update statements to apply the changes to the models.  This is effective but very time consuming if you are attempting to scrub a large amount of data.
MySQL, PostgreSQL and SQLite all support some variation of bulk updating via an upsert operation.  Thankfully `activerecord-import` has incorporated support for this making it easy to implement.  https://github.com/zdennis/activerecord-import#duplicate-key-update
So at the core of this change, we are adding support to apply scrubbed changes via a bulk upsert versus update commands.  
The performance implications of this change are significant.  In a test against a table with 700,000 rows, the upsert approach was approximately **5x faster** (5m13.113s vs 27m49.008s).  
Additional changes in this PR are around reorganizing the code to create one distinct implementation path between `scrub` and `scrub:model` and decomposing some of the layers for better separation of concerns.
Note:  Because the `activerecord-import` implementations for MySQL and PostgreSQL are slightly different, I have only added support for MySQL at this time.

### CODE REVIEW

File: lib/acts_as_scrubbable/ar_class_processor.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    def process
+      # Removing any find or initialize callbacks from model
+        ar_class.reset_callbacks(:initialize)
+        ar_class.reset_callbacks(:find)
+
+        ActsAsScrubbable.logger.info "Scrubbing #{ar_class} ...".green
+
+        num_of_batches = Integer(ENV.fetch("SCRUB_BATCHES", "256")) if num_of_batches.nil?
+        scrubbed_count = ActsAsScrubbable::ParallelTableScrubber.new(ar_class, num_of_batches).scrub do |query|
```
Comment (adamstegman) [inline lib/acts_as_scrubbable/ar_class_processor.rb:29]: With all classes using the parallel table scrubber, is that too many parallel tasks at the same time? Wondering if the overhead is actually hurting us on a bulk scrub.

I'd appreciate some benchmarks on the current version v. this, both for lots of tables and single tables.

File: lib/acts_as_scrubbable/ar_class_processor.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    end
+
+    def process
+      # Removing any find or initialize callbacks from model
+        ar_class.reset_callbacks(:initialize)
+        ar_class.reset_callbacks(:find)
+
+        ActsAsScrubbable.logger.info "Scrubbing #{ar_class} ...".green
+
+        num_of_batches = Integer(ENV.fetch("SCRUB_BATCHES", "256")) if num_of_batches.nil?
```
Comment (adamstegman) [inline lib/acts_as_scrubbable/ar_class_processor.rb:28]: This variable is set here, so no need to check if it's already set.

```suggestion
        num_of_batches = Integer(ENV.fetch("SCRUB_BATCHES", "256"))
```

File: lib/acts_as_scrubbable/ar_class_processor.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        scrubbed_count = ActsAsScrubbable::ParallelTableScrubber.new(ar_class, num_of_batches).scrub do |query|
+          process_class(query)
```
Comment (adamstegman) [inline lib/acts_as_scrubbable/ar_class_processor.rb:30]: This API is a little confusing:

- `scrub` yields a query, and doesn't do any scrubbing, but returns the number of scrubbed objects
- `process_class` is expected to scrub each record and return a count of scrubbed objects

I wonder if there's a way to restructure/rename this that's more straightforward. What about something like this?

```ruby
scrubbed_count = 0
ActsAsScrubbable::ParallelQueries.new(ar_class, num_of_batches).each_query do |query|
  scrubbed_count += scrub_query(query)
end
```

File: lib/acts_as_scrubbable/update_processor.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
+require 'acts_as_scrubbable/base_processor'
+
+module ActsAsScrubbable
+  module UpdateProcessor
+    include BaseProcessor
+
+    def handle_batch(batch)
+      scrubbed_count = 0
+      batch.each do |obj|
+        _updates = obj.scrub!
```
Comment (adamstegman) [inline lib/acts_as_scrubbable/update_processor.rb:10]: This is also confusing—can we rename this method to something like `obj.scrubbed_values`?

```suggestion
        _updates = obj.scrubbed_values
```

File: lib/acts_as_scrubbable.rb [hunk 5]
```diff
@@ -2,16 +2,17 @@
 require 'active_record/version'
 require 'active_support/core_ext/module'
 require 'acts_as_scrubbable/tasks'
-
+require 'term/ansicolor'
+require 'logger'
 
 module ActsAsScrubbable
   extend ActiveSupport::Autoload
+  include Term::ANSIColor
```
[the remark this answers is not in the document]
Comment (mitch-lindsay) [reply lib/acts_as_scrubbable.rb:10]: Should be covered by this line, right?  https://github.com/onemedical/acts_as_scrubbable/blob/update_to_use_import/acts_as_scrubbable.gemspec#L21

[the remark this answers is not in the document]
Comment (adamstegman) [reply lib/acts_as_scrubbable.rb:10]: Doh, I don't know how I missed that!

File: lib/acts_as_scrubbable/ar_class_processor.rb (hunk 1 above)
Comment (mitch-lindsay) [reply lib/acts_as_scrubbable/ar_class_processor.rb]: That's an interesting question.  I've run a couple tests.  Here are the results.

```
WITH UPDATES
> time rake scrub SCRUB_CLASSES=PhoneNumber SKIP_AFTERHOOK=true SKIP_CONFIRM=true
2021-08-09 09:19:15 -0700: [INFO] - Scrubbing PhoneNumber ...
2021-08-09 09:23:48 -0700: [INFO] - 719285 PhoneNumber objects scrubbed
2021-08-09 09:23:48 -0700: [INFO] - Scrub Complete!

real	5m13.113s
user	3m52.744s
sys	0m14.436s

> time rake scrub:model[PhoneNumber] SKIP_CONFIRM=true
2021-08-09 09:25:34 -0700: [INFO] - Scrubbing PhoneNumber ...
2021-08-09 09:26:22 -0700: [INFO] - 719285 PhoneNumber objects scrubbed
2021-08-09 09:26:22 -0700: [INFO] - Scrub Complete!

real	1m12.285s
user	6m37.562s
sys	0m19.508s

> time rake scrub SCRUB_CLASSES=Address,AppointmentInventory,Email,PhoneNumber SKIP_AFTERHOOK=true SKIP_CONFIRM=true
2021-08-09 12:02:55 -0700: [INFO] - Scrubbing Address ...
2021-08-09 12:02:55 -0700: [INFO] - Scrubbing PhoneNumber ...
2021-08-09 12:02:55 -0700: [INFO] - Scrubbing Email ...
2021-08-09 12:02:55 -0700: [INFO] - Scrubbing AppointmentInventory ...
2021-08-09 12:02:57 -0700: [INFO] - 963 Email objects scrubbed
2021-08-09 12:02:57 -0700: [INFO] - Scrub Complete!
2021-08-09 12:03:00 -0700: [INFO] - 829 Address objects scrubbed
2021-08-09 12:03:00 -0700: [INFO] - Scrub Complete!
2021-08-09 12:06:04 -0700: [INFO] - 271799 AppointmentInventory objects scrubbed
2021-08-09 12:06:04 -0700: [INFO] - Scrub Complete!
2021-08-09 12:08:45 -0700: [INFO] - 719285 PhoneNumber objects scrubbed
2021-08-09 12:08:45 -0700: [INFO] - Scrub Complete!

real	6m36.774s
user	7m52.408s
sys	0m23.907s

ORIGINAL
>time rake scrub SCRUB_CLASSES=PhoneNumber SKIP_AFTERHOOK=true SKIP_CONFIRM=true
2021-08-09 10:07:01 -0700: [INFO] - Scrubbable Classes: PhoneNumber
2021-08-09 10:07:01 -0700: [INFO] - Scrubbing PhoneNumber ...
2021-08-09 10:33:53 -0700: [INFO] - 719285 PhoneNumber objects scrubbed
2021-08-09 10:33:53 -0700: [INFO] - Scrub Complete!

real	27m31.251s
user	9m6.228s
sys	0m55.984s

> time rake scrub:model[PhoneNumber] SKIP_CONFIRM=true
2021-08-09 10:37:40 -0700: [INFO] - Scrubbing PhoneNumber ...
2021-08-09 10:42:17 -0700: [INFO] - 719285 PhoneNumber objects scrubbed
2021-08-09 10:42:17 -0700: [INFO] - Scrub Complete!

real	5m14.835s
user	15m19.444s
sys	1m40.528s

> time rake scrub SCRUB_CLASSES=Address,AppointmentInventory,Email,PhoneNumber SKIP_AFTERHOOK=true SKIP_CONFIRM=true
2021-08-09 11:28:47 -0700: [INFO] - Scrubbable Classes: Address, AppointmentInventory, Email, PhoneNumber
2021-08-09 11:28:47 -0700: [INFO] - Scrubbing Address ...
2021-08-09 11:28:47 -0700: [INFO] - Scrubbing AppointmentInventory ...
2021-08-09 11:28:47 -0700: [INFO] - Scrubbing Email ...
2021-08-09 11:28:47 -0700: [INFO] - Scrubbing PhoneNumber ...
2021-08-09 11:28:51 -0700: [INFO] - 963 Email objects scrubbed
2021-08-09 11:28:52 -0700: [INFO] - 829 Address objects scrubbed
2021-08-09 11:30:59 -0700: [INFO] - 271799 AppointmentInventory objects scrubbed
2021-08-09 12:00:53 -0700: [INFO] - 719285 PhoneNumber objects scrubbed
2021-08-09 12:00:53 -0700: [INFO] - Scrub Complete!

real	32m48.553s
user	11m34.385s
sys	1m13.593s
```

File: lib/acts_as_scrubbable/update_processor.rb (hunk 4 above)
Comment (adamstegman) [reply lib/acts_as_scrubbable/update_processor.rb:11]: Resurrecting this thread 😬

File: lib/acts_as_scrubbable/ar_class_processor.rb [hunk 6]
```diff
@@ -3,31 +3,26 @@
...
+      else
+        ActsAsScrubbable.logger.info "Using Update".white
+        @query_processor = UpdateProcessor.new(ar_class)
       end
     end
 
     def process(num_of_batches)
-      # Removing any find or initialize callbacks from model
-      ar_class.reset_callbacks(:initialize)
-      ar_class.reset_callbacks(:find)
```
Comment (adamstegman) [inline lib/acts_as_scrubbable/ar_class_processor.rb]: Should this be moved into the `UpdateProcessor`?

File: lib/acts_as_scrubbable/update_processor.rb (hunk 4 above)
Comment (mitch-lindsay) [reply lib/acts_as_scrubbable/update_processor.rb]: GTG now

**Review by adamstegman — APPROVED** [review]

**Pull request merged**